### PR TITLE
Resource cache tests

### PR DESCRIPTION
**Pull request by ilholmes**
Add tests for PanoptesResourceCache and related changes

### CODE REVIEW

File: yahoo_panoptes/framework/resources.py [hunk 1]
```diff
@@ -868,18 +862,22 @@ def setup_resource_cache(self):
         try:
             self._resources_store = PanoptesResourceStore(self._panoptes_context)
         except Exception as e:
-            raise e
+            logger.error('Error while setting up PanoptesResourceStore: %s' % repr(e))
+            raise PanoptesResourceError("Error while setting up resources_store for PanoptesResourceCache")
```
Comment (varunvarma) [inline yahoo_panoptes/framework/resources.py]: This is a good pattern - however, I would suggest creating a new exception class called PanoptesResourceCacheException for these and dropping the phrase 'for PanoptesResourceCache' in the error descriptions.

File: yahoo_panoptes/framework/resources.py [hunk 2]
```diff
@@ -986,10 +984,13 @@ def close_resource_cache(self):
         Returns:
             None
         """
-        self._db.close()
+        if self._db:
+            self._db.close()
+        else:
+            raise PanoptesResourceError("Attempted to close connection to SQLite db that was not open")
```
Comment (varunvarma) [inline yahoo_panoptes/framework/resources.py]: I have mixed feelings about raising this exception vs. just logging the exception.

File: yahoo_panoptes/framework/resources.py [hunk 3]
```diff
@@ -832,18 +832,12 @@ def get_resources(self, query):
...
-                    metadata_keys = row[8].split('|')
-                    metadata_values = row[9].split('|')
-                    for i in range(0, len(metadata_keys) - 1):
-                        resource.add_metadata(metadata_keys[i], metadata_values[i])
-            except Exception as e:
-                logger.error(
-                        'Either resource metadata key or value are not correct, skipping resource "%s": %s' % (
-                            row, str(e)))
-                continue
+            # Columns 8 & 9 contain the metadata keys and values respectively
```
Comment (varunvarma) [inline yahoo_panoptes/framework/resources.py]: My guess is that this change was made to facilitate testing - however, this skipping of resources which can't be parsed is intentional. We need to revert these changes and the way to test the exception would be to insert a malformed resource directly into the mocked PanoptesResourceKVStore and check the count of fetched resources is one less than expected.

Comment (ilholmes) [reply yahoo_panoptes/framework/resources.py]: Ok, I was ambivalent about removing the try-except, but chose to because I can't figure out how un-parseable metadata could be added to the resource because "add_metadata" has strict checks. Will add this back in on the significant chance I overlooked something.

File: yahoo_panoptes/framework/resources.py (hunk 2 above)
Comment (ilholmes) [reply yahoo_panoptes/framework/resources.py]: I think just logging the exception is the right way to go. Presumably if close_resource_cache is being called, the user has already accomplished what they wanted with it and the in-memory db will be wiped anyway as that process exits.

File: yahoo_panoptes/framework/resources.py [hunk 4]
```diff
@@ -837,12 +844,12 @@ def get_resources(self, query):
                 if row[8] and row[9]:
                     metadata_keys = row[8].split('|')
                     metadata_values = row[9].split('|')
-                    for i in range(0, len(metadata_keys) - 1):
```
Comment (ilholmes) [inline yahoo_panoptes/framework/resources.py:840]: OBOE

File: tests/test_framework.py [hunk 5]
```diff
@@ -452,6 +459,84 @@ def test_resource_dsl_parsing(self):
...
+            mock_get_resources = Mock(side_effect=Exception)
+            with patch('yahoo_panoptes.framework.resources.PanoptesResourceStore.get_resources', mock_get_resources):
+                with self.assertRaises(PanoptesResourceError):
+                    panoptes_resource_cache.setup_resource_cache()
+
+            # Test basic operations
+            panoptes_resource_cache.setup_resource_cache()
+            self.assertIsInstance(panoptes_resource_cache.get_resources('resource_class = "network"'),
+                                  PanoptesResourceSet)
+            self.assertEqual(0, len(panoptes_resource_cache.get_resources('resource_class = "network"')))
```
Comment (varunvarma) [inline tests/test_framework.py]: The pattern we've been using in other tests is assert(returnedValue, expectedValue) - so this should look like: 
self.assertEqual(len(panoptes_resource_cache.get_resources('resource_class = "network"')), 0)

File: yahoo_panoptes/framework/resources.py (hunk 4 above)
Comment (varunvarma) [reply yahoo_panoptes/framework/resources.py:840]: My preferred woodwind instrument is an English Horn, but I can be convinced of looking into an Oboe...

File: tests/test_framework.py [hunk 6]
```diff
@@ -74,6 +75,12 @@ def panoptes_mock_kazoo_client(**kwargs):
     return FakeClient()
 
 
+def mock_fetchall(self):
```
Comment (varunvarma) [inline tests/test_framework.py]: Is this being used anywhere?

File: tests/test_framework.py [hunk 7]
```diff
@@ -452,6 +459,84 @@ def test_resource_dsl_parsing(self):
...
+        serialized_key, serialized_value = PanoptesResourceStore._serialize_resource(panoptes_resource)
+        kv.set(serialized_key, serialized_value)
+
+        mock_kv_store = Mock(return_value=kv)
+
+        with patch('yahoo_panoptes.framework.resources.PanoptesContext.get_kv_store', mock_kv_store):
+            # Test errors when setting up resource cache
+            mock_resource_store = Mock(side_effect=Exception)
+            with patch('yahoo_panoptes.framework.resources.PanoptesResourceStore', mock_resource_store):
+                with self.assertRaises(PanoptesResourceError):
```
Comment (varunvarma) [inline tests/test_framework.py]: Now that we have the PanoptesResourceCacheError class, let's assert for those in these tests

**Review by varunvarma — CHANGES_REQUESTED** [review]
